src/pymoca/tree.py

```diff
@@ -259,6 +290,1052 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+            apply_modification = ast.ClassModification()
+            for arg in instance.modification_environment.arguments:
+                if isinstance(arg.value, ast.ElementModification):
+                    for elem_class_mod in arg.value.modifications:
+                        for sub_arg in elem_class_mod.arguments:
+                            apply_modification.arguments.append(sub_arg)
+                elif isinstance(arg.value, ast.ShortClassDefinition):
+                    apply_modification.arguments.append(arg)
+
+            instance.modification_environment = apply_modification
+        elif element.name not in InstanceTree.BUILTIN_TYPES:
+            # Symbol
+            if ast_ref.class_modification:
+                sym_mod = self._append_modifications(ast_ref.class_modification)
+            else:
+                sym_mod = ast.ClassModification()
+            for arg in instance.modification_environment.arguments:
+                if isinstance(arg.value, ast.ElementModification):
+                    if arg.value.component.indices != [[None]]:
+                        raise ModelicaSemanticError("Subscripting modifiers is not allowed.")
+                    if len(arg.value.component.child):
+                        # Move component reference down a level and apply modification to symbol
+                        # Don't stomp on original that may be used elsewhere
+                        arg = copy.copy(arg)
+                        arg.value = copy.copy(arg.value)
+                        arg.value.component = arg.value.component.child[0]
+                        sym_mod.arguments.append(arg)
+                    else:
+                        for sub_arg in arg.value.modifications:
+                            if isinstance(sub_arg, ast.ClassModification):
+                                sym_mod.arguments += sub_arg.arguments
+                            else:
+                                # Value modification - apply to symbol
+                                # Don't stomp on original that may be used elsewhere
+                                sub_arg = copy.copy(arg)
+                                sub_arg.value = copy.copy(arg.value)
+                                sub_arg.value.component = ast.ComponentRef(name="value")
+                                sym_mod.arguments.append(sub_arg)
+                else:
+                    raise UnimplementedError(f"{arg.value.__class__} symbol modification")
+
+            instance.modification_environment = sym_mod
+
+        # Set modification argument scope now that we have an instance
+        for index, arg in enumerate(instance.modification_environment.arguments):
+            if arg.scope is None:
+                # Make a copy so we don't change original AST or same arg used elsewhere
+                new_arg = copy.copy(arg)
+                new_arg.scope = instance.parent
+                instance.modification_environment.arguments[index] = new_arg
```

Comment:
I think this is often not right. The scope should be the level where the component/symbol is declared, not the parent of the current instance. Modification arguments can be passed multiple levels down the stack in the recursion before they are applied to an instance, but they should retain their original scope from the declaration so that names in the modification values are properly resolved.

This should show up in flattening tests, so I'm going to just add a TODO for now.

Reply:
TODO added in 0f146c4.